services/sidecar/rollup/derivation/batch_v0_encoder.go

```diff
@@ -0,0 +1,169 @@
+	return &BatchV0Encoder{cfg, 0, &bytes.Buffer{}, newSubBatch()}
+}
+
+func (e *BatchV0Encoder) GetBatch(l1Head spTypes.BlockID) ([]byte, error) {
+	if e.timeout != 0 && l1Head.GetNumber() >= e.timeout {
+		data := e.currBuf.Bytes()
+		e.currBuf.Reset()
+		return data, nil
+	}
+	if e.currBuf.Len() < int(e.cfg.GetTargetBatchSize()) {
+		return nil, errBatchTooSmall
+	}
+	data := e.currBuf.Bytes()
+	e.currBuf.Reset()
+	return data, nil
+}
+
+func (e *BatchV0Encoder) Reset() {
+	e.currBuf.Reset()
+	e.subBatch = newSubBatch()
+}
+
+// Processes a block. If the block is non-empty and fits, add it to the current sub-batch.
+// If the block would cause the batch to exceed the target size, close the entire batch (by writing to `currBuf“).
+// Returns an error if the block could not be processed.
+func (e *BatchV0Encoder) ProcessBlock(block *types.Block) error {
+	// Initialize buffer with version byte if it hasn't been already.
+	if e.currBuf.Len() == 0 {
+		if err := e.currBuf.WriteByte(e.getVersion()); err != nil {
+			return fmt.Errorf("failed to encode version: %w", err)
+		}
+	}
+	var (
+		// Block is empty
+		shouldSkipBlock = len(block.Transactions()) == 0
+		// Block would cause the batch to exceed the target size
+		shouldCloseBatch    = uint64(e.currBuf.Len())+e.subBatch.size() > e.cfg.GetTargetBatchSize()
+		shouldCloseSubBatch = shouldCloseBatch || (shouldSkipBlock && e.subBatch.size() > 0)
+	)
+	if shouldCloseSubBatch {
+		if err := e.closeSubBatch(); err != nil {
+			return fmt.Errorf("could not close sub-batch: %w", err)
+		}
+	}
+	// Enforce soft cap on batch size.
+	if shouldCloseBatch {
+		return errBatchFull
+	}
+	// Skip intrinsically-derivable blocks.
+	if shouldSkipBlock {
+		log.Info("Skipping intrinsically-derivable block", "block#", block.NumberU64())
+		return nil
+	}
+	// Decode oracle tx.
+	l1Epoch, _, _, _, _, err := bridge.UnpackL1OracleInput(block.Transactions()[0])
+	if err != nil {
+		return fmt.Errorf("could not unpack oracle tx: %w", err)
+	}
+	e.updateTimeout(l1Epoch.Uint64())
+	// Append a block's txs to the current sub-batch.
+	if err := e.subBatch.appendTxBlock(block.NumberU64(), block.Transactions()); err != nil {
+		return fmt.Errorf("could not append block of txs: %w", err)
+	}
+	return nil
+}
+
+// Returns the data format version (v0, i.e. 0x0).
+func (e *BatchV0Encoder) getVersion() byte { return v0 }
+
+// Updates the batch timeout if the given L1 epoch is earlier than the current timeout.
+// Note: the timeout won't be updated more than once assuming the L1 epoch is monotonically increasing.
+func (e *BatchV0Encoder) updateTimeout(l1Epoch uint64) {
+	timeout := l1Epoch + e.cfg.GetSeqWindowSize() - e.cfg.GetSubSafetyMargin()
+	if e.timeout == 0 || e.timeout > timeout {
+		e.timeout = timeout
+	}
+}
+
+// Writes encoded sub-batch out to the buffer.
+func (e *BatchV0Encoder) closeSubBatch() error {
+	log.Info("Closing sub-batch...")
+	if err := rlp.Encode(e.currBuf, e.subBatch); err != nil {
```

Comment:
this is wrong: appending encoded `subBatch` to `currBuf` does not make it a list of `subBatch`. You should notice that `rlp(subBatch1) ++ rlp(subBatch2)`, `rlp([rlp(subBatch1), rlp(subBatch2)])` and `rlp([subBatch1, subBatch2])` are all different. The last format should be used. Therefore, you should maintain a `[]subBatch` field in the encoder. I'd suggest you create a test that test the encoding and decoding together to make sure it works.